pycoal/mineral.py

```diff
@@ -2,132 +2,197 @@
+WAVELENGTH_MIN = 0.38315
+WAVELENGTH_MAX = 2.5082
+
+class MineralClassification:
+    def __init__(self):
+        # maybe we could put the constants above in here?
+        # any other variables to move in here?
+        pass
+
+    def normalize(self, image, wavelengthList=None):
+        """
+        Return a copy of an MxNxB image (or 1x1xB pixel) with constrained wavelength range and normalized spectra.
+        Args:
+           image (numpy.ndarray): the image or pixel to normalize
+           wavelengthList (list) [default None]: the list of wavelengths for the pixel
+        Returns:
+           normalizedImage (numpy.ndarray): the normalized image or pixel
+        """
+
+        # constrain wavelength range
+        if wavelengthList is None:
+            wavelengthList = image.metadata.get('wavelength')
+
+        # get the indices of the minimum and maximum element
+        i = self.__indexOfGreaterThan(wavelengthList, WAVELENGTH_MIN)
+        j = self.__indexOfGreaterThan(wavelengthList, WAVELENGTH_MAX)
+
+        # constrain wavelength range
+        # slice the array from MxNxB to MxNx(j-i) for indices i,j
+        image = image[:,:,i:j]
+
+        # normalize spectra
+        principalComponents = spectral.algorithms.principal_components(image)
+        principalComponents.reduce(NUMBER_OF_LAYERS)
+        normalizedImage = principalComponents.transform(image)
+
+        # return normalized image
+        return normalizedImage
+
+    def __indexOfGreaterThan(self, elements, value):
+        """
+        Return the index of the first element greater than the value.
+        Args:
+           elements (list): list of elements
+           value (float): the value to compare
+        Returns:
+           index (int): the index of the first element greater than the value
+        """
+
+        for index,element in enumerate(elements):
+            if element > value:
+                return index
+
+    def trainClassifier(self, libraryFileName):
+        """
+        This takes the file name of the library to be used in the training and returns a
+        classifier
+        Args:
+           libraryFileName (str): the name of the training library
+        Returns:
+           classifier (PerceptronClassifier): trained classifier
+        """
+
+        # open the digital spectral library
+        # we may need to load it as an image instead
+        library = spectral.io.envi.open(libraryFileName)
+
+        # get dimensions of library to create mask for training class
+        dimensions = library.spectra.shape
+
+        # create mask
+        mask = [[x for x in range(dimensions[1])] for y in range(dimensions[0])]
+
+        # generate training data
+        trainingData = spectral.algorithms.create_training_classes(self.normalize(library), mask)
+
+        # initialize and train a neural network using the training data
+        classifier = spectral.classifiers.PerceptronClassifier(???)
```

Comment:
Nope, PerceptronClassifier was just there since that was the one we were planning one using, but it does make more sense to make it compatible with the ones you mentioned and not just PerceptronClassifier.